Upstream-Status: Inappropriate [ugly workaround for hailo bug]

---

diff --git a/linux/vdma/memory.c b/linux/vdma/memory.c
index 8c210c4..559f5f1 100644
--- a/linux/vdma/memory.c
+++ b/linux/vdma/memory.c
@@ -167,7 +167,7 @@
     }
 
     if (HAILO_DMA_DMABUF_BUFFER != buffer_type) {
-        vma = find_vma(current->mm, addr_or_fd);
+        vma = find_vma_hailo(current->mm, addr_or_fd);
         if (IS_ENABLED(HAILO_SUPPORT_MMIO_DMA_MAPPING)) {
             if (NULL == vma) {
                 dev_err(dev, "no vma for virt_addr/size = 0x%08lx/0x%08zx\n", addr_or_fd, size);
